src/main/scala/io/iohk/ethereum/domain/Blockchain.scala

```diff
@@ -323,20 +351,64 @@ class BlockchainImpl(
       )
     }
 
+    val checkpointUpdates = maybeBlockHeader match {
+      case Some(header) =>
+        if (header.hasCheckpoint && header.number == getLatestCheckpointBlockNumber()) {
+          val prev = findPreviousCheckpointBlockNumber(header.number, header.number)
+          prev.map { num =>
+            // side effect
+            saveLatestCheckpointNumber(num)
+            appStateStorage.putLatestCheckpointBlockNumber(num)
```

Comment:
I am not sure about this bit.
Lets say we  have saved `X` blocks with all the mpt nodes i.e best checkpoint and best known block is `X`. (for simplicity checkpoint and bestblock are the same)
Now we are on `X + 20` (without updating bestblock and checkpoint in db) , last checkpoint was on `X + 10`.  We start to rollback and and we find this checkpint an we save  into db. Now node dies.
After restart, best checkpoint in db  will be `X + 10` but best block will be `X` as this was last time we sved all data.
Imo if we must to update best checkpoint in db, we should also update best block.